File: libs/py/documind_core/middleware.py
"""
FastAPI middleware stack (Design Areas 62 — Observability, 5 — Tenant,
Global CLAUDE.md §4.3 — Security headers, §4.4 — Rate limiting).

Five middlewares, applied in THIS ORDER at every service::

    CorrelationIdMiddleware      # stamps request with UUID + propagates to logs
    SecurityHeadersMiddleware     # CSP, HSTS, X-Frame-Options, …
    TenantContextMiddleware       # extracts tenant_id from JWT, sets contextvars
    RateLimitMiddleware          # per-tenant + per-IP sliding-window limiter
    ExceptionHandlerMiddleware   # converts AppError → JSON error envelope

Order matters: we want correlation IDs on EVERY log line (even for requests
that error before authentication), so CorrelationId comes first. Rate limit
runs *after* tenant extraction so we can apply tenant-specific limits.

Why not ``BaseHTTPMiddleware``
------------------------------
FastAPI's ``BaseHTTPMiddleware`` has a known performance issue under load
(it consumes the body into memory). We use the raw ASGI ``__call__``
interface for anything in the hot path.
"""
from __future__ import annotations

import logging
import time
import uuid
from collections.abc import Awaitable, Callable

from fastapi import Request, Response
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp

from .exceptions import AppError, RateLimitedError
from .logging_config import bind_request_context, clear_request_context
from .rate_limiter import RateLimiter, tenant_key, ip_key

log = logging.getLogger(__name__)

CORRELATION_HEADER = "X-Correlation-ID"


# ---------------------------------------------------------------------------
# 1. Correlation ID
# ---------------------------------------------------------------------------
class CorrelationIdMiddleware(BaseHTTPMiddleware):
    """
    Reads (or generates) a correlation ID and binds it to:

    * the ContextVar used by structlog (so every log line includes it);
    * the response header (so clients can report it back to support).

    If the caller sends ``X-Correlation-ID``, we trust + propagate it — this
    is how the frontend links its Sentry trace to our server logs.
    """

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        cid = request.headers.get(CORRELATION_HEADER) or str(uuid.uuid4())
        bind_request_context(correlation_id=cid)
        request.state.correlation_id = cid

        start = time.monotonic()
        try:
            response: Response = await call_next(request)
        finally:
            elapsed_ms = (time.monotonic() - start) * 1000
            log.info(
                "request_complete method=%s path=%s duration_ms=%.1f",
                request.method, request.url.path, elapsed_ms,
            )
            clear_request_context()

        response.headers[CORRELATION_HEADER] = cid
        return response


# ---------------------------------------------------------------------------
# 2. Security headers
# ---------------------------------------------------------------------------
class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Add the baseline security headers recommended by OWASP to every response.

    These are CHEAP — a few bytes per response — and fix whole classes of
    vulnerabilities (clickjacking, MIME sniffing, mixed content).
    """

    def __init__(self, app: ASGIApp, csp: str | None = None) -> None:
        super().__init__(app)
        self.csp = csp or "default-src 'self'"

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        response = await call_next(request)
        response.headers.setdefault("X-Content-Type-Options", "nosniff")
        response.headers.setdefault("X-Frame-Options", "DENY")
        response.headers.setdefault("X-XSS-Protection", "1; mode=block")
        response.headers.setdefault("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
        response.headers.setdefault("Content-Security-Policy", self.csp)
        response.headers.setdefault("Referrer-Policy", "strict-origin-when-cross-origin")
        response.headers.setdefault(
            "Permissions-Policy",
            "camera=(), microphone=(), geolocation=()",
        )
        return response


# ---------------------------------------------------------------------------
# 3. Tenant context (extract from Authorization JWT or X-Tenant-ID header)
# ---------------------------------------------------------------------------
class TenantContextMiddleware(BaseHTTPMiddleware):
    """
    Extract the tenant identifier from the request and bind it to the logging
    context.

    In dev, we accept an ``X-Tenant-ID`` header for ease of testing. In
    production, the API Gateway validates the JWT and forwards the tenant in
    a signed header — services here just read that header and bind it.

    Services should still validate that the authenticated principal HAS
    access to the stated tenant — that's the Identity Service's job, done
    at the gateway. This middleware is about LOGGING + RLS context, not
    access control.
    """

    def __init__(self, app: ASGIApp, *, header: str = "X-Tenant-ID", default: str = "") -> None:
        super().__init__(app)
        self.header = header
        self.default = default

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        tenant_id = request.headers.get(self.header, self.default)
        user_id = request.headers.get("X-User-ID", "")
        bind_request_context(
            correlation_id=getattr(request.state, "correlation_id", ""),
            tenant_id=tenant_id,
            user_id=user_id,
        )
        request.state.tenant_id = tenant_id
        request.state.user_id = user_id
        return await call_next(request)


# ---------------------------------------------------------------------------
# 4. Rate limit
# ---------------------------------------------------------------------------
class RateLimitMiddleware(BaseHTTPMiddleware):
    """
    Per-tenant (authenticated) or per-IP (anonymous) sliding-window limiter.

    Path-specific limits (``/api/v1/admin/*`` stricter than ``/api/v1/docs``)
    are handled by the caller wiring multiple middleware instances or by a
    callback that returns the right (limit, window) tuple per request.
    """

    def __init__(
        self,
        app: ASGIApp,
        *,
        limiter: RateLimiter,
        default_limit_per_min: int = 100,
        admin_limit_per_min: int = 50,
        upload_limit_per_min: int = 10,
    ) -> None:
        super().__init__(app)
        self.limiter = limiter
        self.default_limit = default_limit_per_min
        self.admin_limit = admin_limit_per_min
        self.upload_limit = upload_limit_per_min

    def _select_budget(self, path: str, method: str) -> tuple[int, int, str]:
        """Return ``(limit, window_seconds, endpoint_label)`` for the request."""
        if path.startswith("/api/v1/admin"):
            return self.admin_limit, 60, "admin"
        if method == "POST" and "/upload" in path:
            return self.upload_limit, 60, "upload"
        return self.default_limit, 60, "api"

    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        # Skip health + metrics endpoints — they must not be rate-limited
        if request.url.path in ("/health", "/healthz", "/metrics"):
            return await call_next(request)

        tenant_id = getattr(request.state, "tenant_id", "")
        limit, window, endpoint = self._select_budget(request.url.path, request.method)

        key = (
            tenant_key(tenant_id, endpoint)
            if tenant_id
            else ip_key(request.client.host if request.client else "unknown", endpoint)
        )

        try:
            result = await self.limiter.check_or_raise(
                key=key, limit=limit, window_seconds=window
            )
        except RateLimitedError as exc:
            return JSONResponse(
                status_code=429,
                content=exc.to_dict() | {"correlation_id": getattr(request.state, "correlation_id", "")},
                headers={"Retry-After": str(exc.details.get("retry_after_seconds", window))},
            )

        response = await call_next(request)
        # Expose limit state to clients so they can self-throttle
        response.headers["X-RateLimit-Limit"] = str(result.limit)
        response.headers["X-RateLimit-Remaining"] = str(result.remaining)
        response.headers["X-RateLimit-Reset"] = str(result.reset_in_seconds)
        return response


# ---------------------------------------------------------------------------
# 5. Domain-exception → error envelope
# ---------------------------------------------------------------------------
def register_exception_handlers(app) -> None:  # noqa: ANN001 — FastAPI typing pain
    """
    Install a single handler that converts every :class:`AppError` into the
    consistent error envelope mandated by global CLAUDE.md §6.2::

        {
            "detail": "...",
            "error_code": "...",
            "details": {},
            "correlation_id": "..."
        }

    Services call this in their ``app/main.py`` after instantiating ``app``.
    """
    from fastapi import FastAPI

    assert isinstance(app, FastAPI)  # noqa: S101

    @app.exception_handler(AppError)
    async def _handle_app_error(request: Request, exc: AppError) -> JSONResponse:  # noqa: ANN202
        body = exc.to_dict() | {
            "correlation_id": getattr(request.state, "correlation_id", ""),
        }
        log.warning(
            "app_error error_code=%s status=%d message=%s",
            exc.error_code, exc.http_status, exc.message,
        )
        return JSONResponse(status_code=exc.http_status, content=body)

File: services/api-gateway/internal/middleware/middleware.go
// Package middleware holds the gateway's HTTP middleware stack.
//
// Each middleware is one small function that wraps an http.Handler — that's
// the idiomatic Go way. No framework, just chi's router + standard http.
package middleware

import (
	"context"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"
)

type ctxKey string

const (
	CorrelationIDHeader = "X-Correlation-ID"
	TenantIDHeader      = "X-Tenant-ID"
	UserIDHeader        = "X-User-ID"
	RolesHeader         = "X-User-Roles"

	CtxCorrelationID ctxKey = "correlation_id"
	CtxTenantID      ctxKey = "tenant_id"
	CtxUserID        ctxKey = "user_id"
	CtxRoles         ctxKey = "roles"
)

// CorrelationID reads or generates X-Correlation-ID and stashes it in ctx.
func CorrelationID(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		cid := r.Header.Get(CorrelationIDHeader)
		if cid == "" {
			cid = uuid.NewString()
		}
		w.Header().Set(CorrelationIDHeader, cid)
		ctx := context.WithValue(r.Context(), CtxCorrelationID, cid)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// SecurityHeaders matches libs/py/documind_core/middleware.py.
func SecurityHeaders(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		h := w.Header()
		h.Set("X-Content-Type-Options", "nosniff")
		h.Set("X-Frame-Options", "DENY")
		h.Set("X-XSS-Protection", "1; mode=block")
		h.Set("Strict-Transport-Security", "max-age=31536000; includeSubDomains")
		h.Set("Content-Security-Policy", "default-src 'self'")
		h.Set("Referrer-Policy", "strict-origin-when-cross-origin")
		h.Set("Permissions-Policy", "camera=(), microphone=(), geolocation=()")
		next.ServeHTTP(w, r)
	})
}

// Logger emits a structured-ish line per request (real prod wires zap).
func Logger(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		rw := &statusRecorder{ResponseWriter: w, status: 200}
		next.ServeHTTP(rw, r)
		cid, _ := r.Context().Value(CtxCorrelationID).(string)
		log.Printf(
			"request_complete method=%s path=%s status=%d duration_ms=%d correlation_id=%s",
			r.Method, r.URL.Path, rw.status, time.Since(start).Milliseconds(), cid,
		)
	})
}

type statusRecorder struct {
	http.ResponseWriter
	status int
}

func (r *statusRecorder) WriteHeader(s int) { r.status = s; r.ResponseWriter.WriteHeader(s) }

// CORS adds permissive CORS headers for the configured origins.
func CORS(allowed []string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			origin := r.Header.Get("Origin")
			if contains(allowed, origin) {
				w.Header().Set("Access-Control-Allow-Origin", origin)
				w.Header().Set("Access-Control-Allow-Credentials", "true")
				w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Authorization, X-Tenant-ID, X-Correlation-ID")
				w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
			}
			if r.Method == http.MethodOptions {
				w.WriteHeader(http.StatusNoContent)
				return
			}
			next.ServeHTTP(w, r)
		})
	}
}

// JWTAuth is a stub — in production, this verifies the JWT against the
// identity-svc's published public key. See docs/design-areas/22-identity.md.
//
// For the demo, we trust X-Tenant-ID + X-User-ID headers if present.
func JWTAuth(_ string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// Pass tenant + user through to downstream services
			if t := r.Header.Get(TenantIDHeader); t != "" {
				ctx := context.WithValue(r.Context(), CtxTenantID, t)
				r = r.WithContext(ctx)
			}
			next.ServeHTTP(w, r)
		})
	}
}

// RequireRole is a placeholder for RBAC enforcement.
func RequireRole(_ ...string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			// TODO: check JWT claims against required roles; 403 otherwise.
			next.ServeHTTP(w, r)
		})
	}
}

// RateLimit is a placeholder — the Python services implement full
// sliding-window limiting; the gateway's version would talk to the same
// Redis. See libs/py/documind_core/rate_limiter.py for the shape.
func RateLimit(_ string, _ int, _ string) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			next.ServeHTTP(w, r)
		})
	}
}

func contains(list []string, s string) bool {
	for _, x := range list {
		if strings.EqualFold(strings.TrimSpace(x), strings.TrimSpace(s)) {
			return true
		}
	}
	return false
}
